monash-human-power/raspicam
# Move rendering logic to Backend classes

## Description

`overlay.py` was getting veeery crowded, with lots and lots of conditions for whether or not we were running on the Pi. On the Pi, we use the `picamera` library to capture, record and show video, and on computers, `OpenCV` is used. So, to make things neater, I've made a `Backend` base class from which `PiCameraBackend` and `OpenCVBackend` derive. The `Backend` class contains methods for starting/stopping video, updating overlays, starting/stopping recordings and sending recording statuses. Some logic is shared (such as how the name of the next recording should be found) but most is very specific to the backend. `overlay.py` can now call methods on whichever `Backend` is loaded, without worrying about what device it is running on.

Other minor changes:
- Moved `Canvas` to its own file - It's no more relevant to `Overlay` as the backend stuff, and also helps clean up the very cluttered `overlay.py` file. It's also used by the backend files, and I didn't like the circular imports between `overlay.py` <--> `backend*.py`.

I don't have my laptop with me right now (long story), so note that at this point this is **untested using the OpenCV backend**. I expect that it won't run, so your first review of this PR will probably take 3 minutes max. Let me know what error it spits out and I'll fix it up as soon as I can :)

## Screenshots

Overlay working (on Raspberry PI) as before.

![image](https://user-images.githubusercontent.com/17876556/83727753-ebcc0180-a688-11ea-9547-77cf18d42cc0.png)

## Steps to Test

Basically, just check everything is working as it did before. `poetry shell`, run the overlays (`python overlay*.py`), check data is displayed on the overlays correctly, check that recording still works (see #33 for instructions).
